### ISSUE TITLE
Enhancement: Programming keypad

### ISSUE THREAD

**Issue by BugZappr** [Qalculate/qalculate-gtk#93]
Other calculators often have a programmer's keypad, such as M$ Calculator.  It seems that all these functions are supported if you type things in via the keyboard and/or choose a different display format, and there is a button with some logical operators and base conversion; but none of these really cater to the programmer.

I would suggest that to really do this right, you would:

- Add a hexadecimal keypad; that included the letters A-F
- Have all the bitwise operators on their own buttons
- Have two-input operators act like other two-input operators; such that you don't click bitwise exclusive OR, and a dialog box comes up.  Instead, you input 18FF, click e.g. a bitwise EOR button, input FFFF, and equals; to get E700.
- When in hexadecimal mode, you don't have to type 0x in front of hexadecimal values in order to enter them.  You don't see 0x in front of your answer.  Similar considerations would apply to octal and binary modes.
- You could switch bases with a single button press instead of a menu
- Bitwise exclusive or might be "^", with the exponent, as always, being expressed by "**"

Note that these modes would change the way text in the upper box is evaluated; as the assigned bases are assumed.

While we are used to typing 0x18FF for the number in source code, calculators in programmer mode don't generally make one type 0x.

**Comment by hanna-kn** [comment]
> Have two-input operators act like other two-input operators; such that you don't click bitwise exclusive OR, and a dialog box comes up. Instead, you input 18FF, click e.g. a bitwise EOR button, input FFFF, and equals; to get E700.

This would be accomplished by the addition of an operator for exclusive OR. In the normal mode '^' is clearly not an appropriate candidate.

> When in hexadecimal mode, you don't have to type 0x in front of hexadecimal values in order to enter them. 

This can currently be accomplished by changing the input base (changed in a dialog, separately from the output base), or using the number bases dialog (which supports calculation of full expressions).

I think an extended version of the base conversion dialog might be most convenient for a programmer's mode. Roman numerals and duodecimal bases should probably be hidden, some custom parsing implemented, and a keypad (which the user should be able to hide - I would expect most programmers to use the keyboard instead of the mouse for input) with 0-F and bitwise (and standard) operators. Most of the functionality of the rest of the GUI (completion, units, conversion, etc.) would probably not be useful and continuous calculation is convenient. (But, if the programmers mode is separated from the rest of the GUI, it might be better with a separate application, which would not really need the functionality of libqalculate, and already has been implemented by others...)

**Comment by hanna-kn** [comment]
Now I have implemented the following:
- A keypad (includes numbers 0-F and bitwise operators) has been added to the number bases window.
- A bitwise XOR (⊻) operator has been added. This can be input using Ctrl+^ or, if set in preferences, '^' on the keyboard. All bitwise operators has been added to the number bases button menu.
- Base prefixes are now by default removed from the displayed result.

**Comment by BugZappr** [comment]
I would suggest that the preferences option only affect the programmer's mode.  Thus, when one goes back to normal mode (one I would call scientific mode), ^ for exponent operates normally.  That way, it would cater to programmers in programmer's mode; but not in other modes.  I don't really follow what you mean about a number bases window.  I see a number bases menu choice, but not a menu for it.  I suppose what I'm really asking for, would be well-implemented by having a choice in addition to "Keypad", "History" and "Conversion", perhaps called "Programmer's Mode", that would use the currently defined base, not just for display, but input.  It would also interpret formulas specially, so as e.g. to recognize all "C" operators, including bitwise and shift operators (while we're at it).  I wouldn't expect it to mirror "C" math functions, but ** would work on real numbers like normal for Qalculate.  (If we were being completely rigorous here, e.g. hexadecimal values could have hex digits to the right of a decimal point, too, but then that would be outside the scope of a programmer's mode.) When the history loads an expression, and the expression was entered in the Programming mode, it might switch to it.  After some time of using Qalculate, and being deeply impressed, this strikes me as the biggest thing other calculators have, that this one doesn't, yet.

**Comment by hanna-kn** [comment]
Which C operators are you missing? 
I will "bitor", "bitand", and "compl" as alternatives to '|', '&', and '~', which are already supported (as is "<<" and ">>" for bitwise shift). 

These are the changes I've made to the GUI:
![bitwise operators menu](https://user-images.githubusercontent.com/19703699/56902350-65fb8100-6a9a-11e9-88b5-1e7f9a1f980f.png)
![number bases keypad](https://user-images.githubusercontent.com/19703699/56902373-77dd2400-6a9a-11e9-8670-79671bfb2f38.png)

It would be possible to add a similar keypad to the main window (bit I would preferably like to avoid it).

Is there any additional buttons you think should be added to the above keypad?

**Comment by BugZappr** [comment]
I think one's compliment is just an exclusive OR; but two's compliment would be good.  Other than that, it looks good to me.  It's good to have a number pad in base conversion, but it's not like having the "C" operators in a calculator right there with a hex keypad, and not having to type 0xFF18, e.g.  Also, I would hope for history in a programmer's calculator.  That's why it seems likely that you should interpret formulas one way for a programming calculator, and another for the remaining 85% as you have it - whether as a pop-up calculator, or a special mode and interpretation of text, of the main window.   Then again, you could have options like: "Interpret '^' as EOR if the hex keypad is displayed"; and "Assume the prefixes for bases when the hex keypad is displayed."   In the command line qalculate, that could be set with a command line switch, stored in ".qalculate"; or just have them edit ".qalculate".  I know modeless programming was supposed to result in a lot more freedom; but in this case, it would result in fewer keystrokes.

**Comment by hanna-kn** [comment]
It seems as if the first priority should be to come up with a way to more easily/efficiently change the input base. The easiest option would be to add keyboard shortcuts. It would also be easy to add base selection to the context menu of the expression entry, but optimally there should be something similar to the new base selection in the number bases keypad.

It should be possible to change the keypad in the main window when another input base is selected.

I would prefer not to add a second keypad to main window, but creating something similar to the number bases window in a separate number bases tab might be feasible (although it would not allow continuous calculation as you type).

An option I might consider is to provide a few different sets of buttons for the left side of the keypad in the main window.

Adding history to the number bases window might be a step too far (but I will nevertheless try to come up with an appropriate design).

In qalc `set caret as xor` or `set XOR^` will (with the recent changes) change the interpretation of '^'.

> However, I'd hope that choosing e.g. Bitwise Shift wouldn't pull up a dialog box that could insert e.g. "shift(4; 3)" when you chose insert, but only "<<" without a dialog box or choosing insert.

This has already been implemented ("<<" has been supported for a very long time).

> "%" for a modulo operator might be a very good example of where it would have a special meaning in a programmer's mode. I'm not sure what it might be like to support "4 % 5" and "4 - 5 % - 3" in the same mode (and I think "1.99 + 9%" is very handy), but it might be possible.

It should be possible to interpret % as modulo if preceded and followed by an integer.

I still do not think that a complete programmers mode, with modified parsing, should be necessary (although integer division using '/' will probably not possible).

**Comment by BugZappr** [comment]
Keyboard shortcuts would make typing 0x into one keystroke, instead of two?  I'm hoping for none.  As to <<, I do see that the text edit section only requires <<.  However, to interact with the buttons, it requires a drop-down menu, and you can't use the buttons to type 7F8 % 4FF.  Once you have 7F8 typed in, and you go to use the modulo button, get a dialog box that conveniently has 7F8 entered, and you type your denominator, and then you type insert, you don't get 7F8 % 3, but "0x7f8mod(0x7f8; 3)".  I think configurable buttons would go a long way, and a mode where you don't have to type base prefixes would too, and being able to have "^" mean EOR would; but a true programming mode, IMHO, would have all three with a single selection and then you could then (in the GUI, anyway) go right back into normal scientific mode.

Note that "4 - 5 % + 3" is ambiguous.  Is it signifying "4 - 5% + 3" (percent) or "4 - ( 5 % 3)" (modulo)?  In a programmer's calculator, modulo could always be assumed.  I was thinking in the general/scientific mode, there could be a rule regarding %, such that if surrounded by spaces, modulo would be preferred; but if following an integer, percent would be preferred.  There is some precedent in "C" for interpreting an operator differently if it is surrounded by spaces.  

In short, programmers need their own mode of interpreting text formulas in the main calculator, or a very similar pop-up one; IMHO.

**Comment by hanna-kn** [comment]
You're misunderstanding. The only possible alternative to using "0x" (or the hex function) is to explicitly change to base of the whole expression. This would not change with a programmers mode. It is possible to guess the base of for example "5f3", but not the hexadecimal number "10".

You can currently change the base (to any base between 2 and 32) of the whole expression, but this requires opening of a dialog, which is less than optimal.
![input base selection](https://user-images.githubusercontent.com/19703699/56941406-a0096900-6b14-11e9-95a0-55d3876d94b2.png)

> Note that "4 - 5 % + 3" is ambiguous. Is it signifying "4 - 5% + 3" (percent) or "4 - ( 5 % 3)" (modulo)?

I think it would be safe to assume that few would type "5 % + 3" instead of "5 % 3" if modulo is intended. "5 % -3" is a bit more tricky, but should be manageable.

**Comment by BugZappr** [comment]
OIC.  You do have the expression base.  In a programming mode then, typically calculators have a single selection of base on the main display, and a single click changes both.

**Comment by hanna-kn** [comment]
A possible programmers keypad:
![programmers keypad](https://user-images.githubusercontent.com/19703699/56944809-19a95300-6b25-11e9-8c9d-dd83cce24a75.png)

`qalc -s "inbase 16" -b 16 18FF xor FFF` currently works.

**Comment by BugZappr** [comment]
This looks outstanding; although I'd like to see the "C" operators instead of the unusual symbols.  It looks like a programmer's mode for qalc may be just a tweak of what you already have.  A flag "`-p16`" (or what makes the most sense to you, but short) might then be shorthand for  "`-s "inbase 16" -b 16`" and special parsing for "C" operators.

**Comment by hanna-kn** [comment]
Mostly done now.

% is interpreted as modulo (rem) operator if not at the beginning or end of expression, and not preceded or followed by another operator (spaces are ignored).

// is now interpreted as operator for integer division (truncated). I thinking of adding support for '\' when followed by space or number (it is also used for variables).

Integer division should probably be added to the keypad (which button should be replaced?). Perhaps the factorize button can be replaced (the menu items might be enough). Is int and frac necessary? Is factorials often useful for programmers? (Replace one with exp()?)

I do not know if logical operators is useful enough for adding GUI elements (and currently negative values are interpreted as false). I would prefer to avoid adding drop-down menus.

The modulo button is problematic. "mod" is misleading as % uses rem() (this is also in the way of supporting " mod " as % alternative). % is label might be confusing, and "rem" might not be as recognizable for programmers. 

The "C" operators are ambiguous as symbols. The unambiguous symbols is also used by gnome-calculator. An alternative would be to use AND, OR, XOR as labels, but AND and OR are used as logical operators.

I agree that the close icon is problematic.

qalc flag has not (yet) been added but `qalc -s "inbase 16" -s "XOR^" -b 16 18FF ^ FFF` works.

**Comment by BugZappr** [comment]
> % is interpreted as modulo (rem) operator if not at the beginning or end of expression, and not preceded or followed by another operator (spaces are ignored).

I can't really say how well that works on the above examples; but I admit that using +3 or -3 as denominator is rare; and most users would soon recognize the problem and revise their entries.

> //" is now interpreted as operator for integer division (truncated). I'm thinking of adding support for '\' when followed by space or number (it is also used for variables).

OK.  Why not.  Of course, that is a comment sequence in C++; but I don't see the harm.

> Integer division should probably be added to the keypad (which button should be replaced?). Perhaps the factorize button can be replaced (the menu items might be enough). Is int and frac necessary? Is factorials often useful for programmers? (Replace one with exp()?)

Most programmers aren't higher-math oriented; so factorization is probably a rarity for them.  Int (i.e. "floor()") and frac (i.e. "trunc()") are indeed used by programmers a fair amount.  Probably not factorials as much.  Yes, "exp()", i.e. e to the power of _x_, is; mostly for scientific work, being the inverse of ln(), asI'm sure you know.

> I do not know if logical operators is useful enough for adding GUI elements (and currently negative values are interpreted as false). I would prefer to avoid adding drop-down menus (on the other hand, it might be nice to also have functions for character codes, timestamps etc. on the keypad). There should be space for another row of buttons without increasing the minimum height of the window.

I myself was resisting the idea of drop-downs for programmer's mode at first; but just so long as the main buttons are single-click as on other programming modes, and the things on the drop-downs are programming oriented, I figure it could only enhance its usability.  I say, let your creativity run wild.

>The modulo button is problematic. "mod" is misleading as % uses rem() (this is also in the way of supporting " mod " as % alternative). % as label might be confusing, and "rem" might not be as recognizable for programmers.

I would expect a "mod" button, as on many scientific calculators.  I would expect it to simply insert "%" in the text area.

> The "C" operators are ambiguous as symbols. The unambiguous symbols is also used by gnome-calculator. An alternative would be to use AND, OR, XOR as labels, but AND and OR are used as logical operators.

THESE might make a good three way option: "Symbols for bitwise operators in programming mode:" and the choices might be 1) "Gnome symbols" (although they may be from a branch of mathematics for bitwise operators), 2) "C" operators, 3) BAND, BOR, BXOR for bitwise versions, and AND, OR, and XOR for logical versions.

> I agree that the close icon is problematic.

Ah.  If that's a close icon, having it in red would help communicate what it is.

> qalc flag has not (yet) been added but qalc -s "inbase 16" -s "XOR^" -b 16 18FF ^ FFF works.

I think they would be used, for the various programmer's bases: decimal, hex, octal, and binary; as convenience macro flags.

I find it refreshing and engaging that you aren't of a fixed, parochial mindset, but are able to handle such a paradigm-shifting enhancement gracefully.  B}

**Comment by hanna-kn** [comment]
I've added another line to be able to fit buttons for exp function (particularly useful since e is not supported in hexadecimal mode), integer division, Unicode conversion (I updated the char() and code() functions), and Unix time stamp conversion (will probably be replaced). 

> I can't really say how well that works on the above examples; but I admit that using +3 or -3 as denominator is rare; and most users would soon recognize the problem and revise their entries.

And % with negative numerator or denominator gives different result in different programming languages.

> OK. Why not. Of course, that is a comment sequence in C++; but I don't see the harm.

// is actually used for integer division in Python. \ (supported when followed by space, digit, left parenthesis, plus, or minus) and "div" (now added) is often used to denote integer division.

> I would expect a "mod" button, as on many scientific calculators. I would expect it to simply insert "%" in the text area.

The "mod" button will now insert the less ambiguous " mod " operator (unless expression is empty). The % operator will calculate remainder as in C.

> Ah. If that's a close icon, having it in red would help communicate what it is. I take it that that just takes you back to general/scientific mode.

I removed the close icon and settled for a keypad context menu, a menu item and a keyboard shortcut. If more keypad variants are added I will probably add buttons for switching keypad (I could not come up with a good icon or text abbreviation for a programming keypad toggle button).

> I myself was resisting the idea of drop-downs for programmer's mode at first; but just so long as the main buttons are single-click as on other programming modes, and the things on the drop-downs are programming oriented, I figure it could only enhance its usability. I say, let your creativity run wild.

It would be difficult to fill all the drop-down menus with enough meaningful content. I'm considering adding middle-click actions instead (already added to the base selection buttons, for partial base changes).

qalc "-p BASE" (same as qalc -b "BASE BASE" -s "xor^") command line option has now been added.

Word size option should probably be added.

> I find it refreshing and engaging (not to mention, impressive) that you aren't of a fixed, parochial mindset, but are able to handle such a paradigm-shifting enhancement gracefully. B}

Thanks. I've really had a lot of fun implementing this request. :smiley:

**Comment by BugZappr** [comment]
No need to fill drop-down menus to have them, IMHO, FWIW.  If only a handful of buttons have them, I don't think it would be odd, for a specialized button pad; and would help people find the text equivalents for these functions.

"mod" makes sense, as long as you can just click buttons for ABCD, then mod, then 3, i.e. a binary operator between the numbers, without an intervening dialog.  Much more like calculators that way.  With touch tablets these days, I'll be very intuitive to those used to it.

I look forward to downing and trying out the new version; but it'll likely take some time for me to get to really trying it out.  I know a lot of peepz are going to love it.

**Comment by BugZappr** [comment]
As the PC version got updated, I've had a chance to try out the new programming keypad myself.  I think its marvelous, and beyond my expectations - and done remarkably quickly.  As you've done such an exceptional job, I hesitate to mention it, but two more tweaks come to mind.  They aren't critical, and I'm sure programmers are going to love the new keypad even without them; but here goes.

1. On most other programming calculators, when you change between bases by pressing Dec, Hex, etc., the number on the main display usually changes to be in the new base.  Currently, if you have a hex number displayed, the calculator simply decides that alphabetics in the number are extraneous.  

2. When an equation is saved in the history and recalled, the base being used isn't remembered; and thus, it might generate an error.  I suggest remembering the number base in the history, such that if a calculation in the history is done in a certain base, the user changes base, and then recalls the history item, its number base be saved, so that the original calculation is instantly recalled upon hitting equals.  One of the pitfalls of modal programming, but a rare case where it's helpful, IMHO.

Random musings: What's the significance of the semicolon in curly brackets, I'm just curious?  You might make the Keypad item a non-editable combo box, and have choices of None, Standard, and Programming.  Should you feel motivated to allow a different one, it could go there.

**Comment by BugZappr** [comment]
My thought, was to change the base of the numbers in the expression when the base is changed.  If in Hex mode, you type FFFF, and change to Dec, it would change to 65535.  This is the way most such programming modes work.  If it was FFFF+sin(FE), it would change to 65535+sin(254), for the sake of consistency.; and so the formulas would remain valid no matter what base the numbers were represented in.  FWIW.  Awesome work, either way 8}

**Comment by cousteaulecommandant** [comment]
Maybe I'm late to the party, but [Sage](https://doc.sagemath.org/html/en/faq/faq-usage.html#how-do-i-use-the-bitwise-xor-operator-in-sage) uses `^` (or `**`) for exponentiation and `^^` for bitwise XOR.  Maybe Qalculate could do the same and use `^^` as an alternative spelling for `⊻`?  It is probably easier to type than Ctrl-^, and I think it's intuitive enough.

Another option is to do like [Lua](https://www.lua.org/manual/5.3/manual.html#3.4.2), which uses `~` for _both_ bitwise not (`~x`) and bitwise xor (`x~y`), although this might be less intuitive.
